misk-core/src/main/kotlin/misk/queuing/TimedBlockingQueue.kt

```diff
@@ -0,0 +1,136 @@
+package misk.queuing
+
+import java.util.concurrent.ArrayBlockingQueue
+import java.util.concurrent.BlockingQueue
+import java.util.concurrent.TimeUnit
+
+/**
+ * This BlockingQueue implementation allows measuring how long dequeued items have been
+ * in the queue. It requires a delayHandler which consumes the latency every time an item is
+ * removed from the queue by {@link BlockingQueue#take()},
+ * {@link BlockingQueue#poll(long, TimeUnit)},
+ * {@link BlockingQueue#poll()} or
+ * {@link BlockingQueue#remove()}.
+ * The operations
+ * {@link BlockingQueue#drainTo(Collection)} and
+ * {@link BlockingQueue#drainTo(Collection, int)} are not supported
+ * @param <T> The type of the elements held in this queue
+ */
+class TimedBlockingQueue<T>(
+  maxQueueSize: Int,
+  private val delayHandler: (Long) -> Unit
+) : BlockingQueue<T> {
+  private val queue: BlockingQueue<TimedQueueItem<T>> =
+      ArrayBlockingQueue<TimedQueueItem<T>>(maxQueueSize)
+
+  override fun poll(timeout: Long, unit: TimeUnit): T? {
+    val item = queue.poll(timeout, unit)
+    return if (item != null) handleRemoved(item) else null
+  }
+
+  override fun poll(): T? {
+    val item = queue.poll()
+    return if (item != null) handleRemoved(item) else null
+  }
+
+  override fun take(): T {
+    return handleRemoved(queue.take())
+  }
+
+  override fun remove(): T {
+    return handleRemoved(queue.remove())
+  }
+
+  override fun contains(element: T): Boolean {
+    return queue.contains(wrap(element))
+  }
+
+  override fun addAll(elements: Collection<T>): Boolean {
+    return queue.addAll(wrapCollection(elements))
+  }
+
+  override fun clear() {
+    queue.clear()
+  }
+
+  override fun element(): T {
+    return unwrap(queue.element())
+  }
+
+  override fun removeAll(elements: Collection<T>): Boolean {
+    return queue.removeAll(wrapCollection(elements))
```

Comment:
shouldn't this invoke the `delayHandler`?

Reply:
The problem is that there is no way to get the delay since these remove operations do not return the queued value (this one returns a boolean).

Reply:
ah right, well that's lame. That would be really surprising to me as a caller, that some methods that remove elements don't update the metrics. I do see the javadoc makes it clear only a few methods actually update metrics, which is probably fine. An alternative is to just not support these other methods?

Reply:
I am working on a workaround to find all matching items before calling `queue.removeAll()` so that we can invoke the delayHandler on each of them.

Reply:
All operations that remove should now invoke the `delayHandler` in the latest commit